src/fparser/two/symbol_table.py

```diff
@@ -0,0 +1,402 @@
+                     associated module or routine.
+    :param parent: the symbol table within which this one is nested (if any).
+    :type parent: :py:class:`fparser.two.symbol_table.SymbolTable.Symbol`
+
+    '''
+    # TODO #201 add support for other symbol properties (kind, shape
+    # and visibility). We may need a distinct Symbol class so as to provide
+    # type checking for the various properties.
+    Symbol = namedtuple("Symbol", "name primitive_type")
+
+    def __init__(self, name, parent=None):
+        self._name = name.lower()
+        # Symbols defined in this scope.
+        self._symbols = {}
+        # Modules imported into this scope.
+        self._modules = {}
+        # Reference to a SymbolTable that contains this one (if any). Actual
+        # value (if any) is set via setter method.
+        self._parent = None
+        self.parent = parent
+        # Symbol tables nested within this one.
+        self._children = []
+
+    def __str__(self):
+        header = "===========\n"
+        symbols = "Symbols:\n"
+        if self._symbols:
+            symbols += "\n".join(list(self._symbols.keys())) + "\n"
+        uses = "Used modules:\n"
+        if self._modules:
+            uses += "\n".join(list(self._modules.keys())) + "\n"
+        return ("{0}Symbol Table '{1}'\n".format(header, self._name) +
+                symbols + uses + header)
+
+    def add_symbol(self, name, primitive_type):
+        '''
+        Creates a new Symbol with the specified properties and adds it to
+        the symbol table. The supplied name is converted to lower case.
+
+        TODO #201 add support for other symbol properties (kind, shape
+        and visibility).
+
+        :param str name: the name of the symbol.
+        :param str primitive_type: the primitive type of the symbol.
+
+        :raises TypeError: if any of the supplied parameters are of the \
+                           wrong type.
+        :raises SymbolTableError: if the symbol table already contains an
+                                  entry with the supplied name.
+        '''
+        if not isinstance(name, six.string_types):
+            raise TypeError("The name of the symbol must be a str but got "
+                            "'{0}'".format(type(name).__name__))
+        # TODO #201 use an enumeration for the primitive type
+        if not isinstance(primitive_type, six.string_types):
+            raise TypeError(
+                "The primitive type of the symbol must be specified as a str "
+                "but got '{0}'".format(type(primitive_type).__name__))
+        lname = name.lower()
+        if lname in self._symbols:
+            raise SymbolTableError("Symbol table already contains an entry "
+                                   "with name '{0}'".format(name))
+        self._symbols[lname] = SymbolTable.Symbol(lname,
+                                                  primitive_type.lower())
+
+    def add_use(self, name, only_list=None):
```

Comment:
Perhaps this should be called `add_use_symbols`

Reply:
Yes, that's better.